software/apps/audio_module/event_detector/main.c

```diff
@@ -0,0 +1,161 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <adc.h>
+#include <console.h>
+#include <gpio.h>
+#include <led.h>
+#include <timer.h>
+#include <tock.h>
+
+#include "app_watchdog.h"
+#include "signpost_api.h"
+#include "simple_post.h"
+#include "_kiss_fft_guts.h"
+#include "kiss_fftr.h"
+#include "common.h"
+#include "ridgeTracker.h"
+
+static int freqAnalyze(kiss_fft_scalar *frame, kiss_fft_scalar *spec);
+static int printMat(char *filename, kiss_fft_scalar *in, size_t N);
+static int printMat2(char *filename, size_t *in, size_t N);
+
+int main (void) {
+    int err = SUCCESS;
+    printf("[Audio Module] Event Detector\n");
+
+    // initialize ADC
+    err = adc_initialize();
+    if (err < SUCCESS) {
+        printf("Initialize errored: %d\n", err);
+    }
+
+    static kiss_fft_scalar frame[BUF_LEN];
+    static kiss_fft_scalar spec[FRE_LEN];
+    static kiss_fft_scalar snrOut[FRE_LEN];
+    memset(frame, 0, BUF_LEN*sizeof(kiss_fft_scalar));
+    size_t frameIdx = 0;
+    size_t globTI = 0;
+    ridgeTracker_init();
+    while (true) {
+        // read data from ADC
+        for (size_t k = frameIdx; k < frameIdx+INC_LEN; k++){
+            err = adc_read_single_sample(3);
+            if (err < SUCCESS) {
+                printf("ADC read error: %d\n", err);
+            }
+            uint16_t sample = err & 0xFFFF;
+            frame[k] = (kiss_fft_scalar)sample;
+        }
+        frameIdx = (frameIdx+INC_LEN) % BUF_LEN;
+
+        // perform FFT
+        if (freqAnalyze(frame, spec)){
+            printf("freqAnalyze() failed!\n");
+            return 1;
+        }
+
+        // ridgetracker update
+        ridgeTracker_update(spec, snrOut);
+
+        // check for available event
+        if (ridgeTracker_isReady){
+            printf("Event detected!\n");
+            printf("ridgeTracker_out.used =  %zu\n",ridgeTracker_out.used);
+
+            if (printMat("SNR.mat",ridgeTracker_out.SNR,ridgeTracker_out.used)){
+                printf("printMat() failed!\n");
+                return 1;
+            }
+            if (printMat2("FI.mat",ridgeTracker_out.FI,ridgeTracker_out.used)){
+                printf("printMat2() failed!\n");
+                return 1;
+            }
+            size_t TIE = ridgeTracker_out.TI[ridgeTracker_out.used-1];
+            for (size_t k=0; k<ridgeTracker_out.used; k++){
+                ridgeTracker_out.TI[k] = globTI-(TIE-ridgeTracker_out.TI[k]);
+            }
+            if (printMat2("TI.mat",ridgeTracker_out.TI,ridgeTracker_out.used)){
+                printf("printMat2() failed!\n");
+                return 1;
+            }
+
+            ridgeTracker_reset();
+        }
+        globTI += 1;
+    }
+
+    ridgeTracker_destroy();
+
+    printf("Done!\n");
+    fflush(stdout);
+    return 0;
+}
+
+static int freqAnalyze(kiss_fft_scalar *frame, kiss_fft_scalar *spec){
```

Comment:
In general, these static helper methods are returning positive integers on error, which is pretty unintuitive for C. I think it would be better to return negative numbers (common for errors) and check `if(helper_function() < 0)` in `main`.